src/statemachine.rs

```diff
@@ -506,47 +460,89 @@ impl StateMachineCtap2 {
+                            },
+                        );
+                        // The DeviceSelector could already be dead, but it might also wait
+                        // for us to respond, in order to cancel all other tokens in case
+                        // we skipped the "blinking"-action and went straight for the actual
+                        // request.
+                        let _ = selector.send(DeviceSelectorEvent::SelectedToken(dev.id()));
                     }
+                    match resp {
+                        Ok(MakeCredentialsResult::CTAP2(attestation, client_data)) => {
+                            callback.call(Ok(RegisterResult::CTAP2(attestation, client_data)))
+                        }
+                        Ok(MakeCredentialsResult::CTAP1(data)) => {
+                            callback.call(Ok(RegisterResult::CTAP1(data, dev.get_device_info())))
+                        }
 
-                    Err(HIDError::Command(CommandError::StatusCode(
-                        StatusCode::ChannelBusy,
-                        _,
-                    ))) => {}
-                    Err(e) => {
-                        warn!("error happened: {}", e);
-                        callback.call(Err(AuthenticatorError::HIDError(e)));
+                        Err(HIDError::Command(CommandError::StatusCode(
+                            StatusCode::ChannelBusy,
+                            _,
+                        ))) => {}
+                        Err(e) => {
+                            match repackage_pin_errors(&mut dev, e.into()) {
+                                AuthenticatorError::PinError(PinError::PinAuthInvalid)
+                                    if pin_uv_auth_result == PinUvAuthResult::UsingInternalUv =>
+                                {
+                                    // This should only happen for CTAP2.0 tokens that use internal UV and failed
+                                    // (e.g. wrong fingerprint used)
+                                    send_status(&status, crate::StatusUpdate::PinAuthInvalid);
+                                    continue;
+                                }
+                                AuthenticatorError::PinError(PinError::PinRequired)
+                                    if pin_uv_auth_result == PinUvAuthResult::UsingInternalUv =>
+                                {
+                                    // This should only happen for CTAP2.0 tokens that use internal UV and failed
+                                    // repeatedly, so that we have to fall back to PINs
+                                    skip_uv = true;
+                                    continue;
+                                }
+                                e => {
+                                    warn!("error happened: {}", e);
+                                    callback.call(Err(e));
+                                }
+                            }
+                        }
```

Comment:
Sadly, they do NOT differ unnecessarily. They actually are different, because the spec (or at least how vendors interpret it) are different for GetAssertion and MakeCredential, in what they return in case of errors.

Reply:
Can't we just handle the difference in `repackage_pin_errors`? If not, I think we should avoid using `repackage_pin_errors` entirely. A flat list of cases would be easier to follow.

Reply:
Since we need `repackage_pin_errors` in other places as well, I've introduced a new function (which also uses it), that handles the differences, so we can move all of it out of `statemachine.rs`